### DISCUSSION TITLE
DIN2USB MIDI Bridging isn't working for me - looking for ideas

### DISCUSSION THREAD

**Discussion by diyelectromusic** [lathoub/Arduino-USBMIDI#15] [Q&A]
Hello!

I'm basically using the MIDI_DIN2USB example with a Pro Micro (ATmega32U4), but find that the USB only seems to send the following to the PC when I press a note on a serial MIDI DIN keyboard (output from MIDIOx for a single keypress and release):
 TIMESTAMP IN PORT STATUS DATA1 DATA2 CHAN NOTE EVENT               
 0000206F   1  --     90    00    00    1  C -1 Note Off              
 000020F8   1  --     80    00    00    1  C -1 Note Off              

If I add in a call to MIDICoreUSB.sendNoteOn directly, it works fine... e.g.

void onSerialMessage(const MidiMessage& message) {
  MIDICoreUSB.send(message);
  switch (message.type) {
    case NoteOn:
      MIDICoreUSB.sendNoteOn(message.data1, message.data2, message.channel);
      break;
    case NoteOff:
      MIDICoreUSB.sendNoteOff(message.data1, message.data2, message.channel);
      break;    
  }
}

This results in the following (again this is all for one keypress and release):
 TIMESTAMP IN PORT STATUS DATA1 DATA2 CHAN NOTE EVENT               
 0000212B   1  --     90    00    00    1  C -1 Note Off              
 0000212B   1  --     90    48    7F    1  C  5 Note On               
 000025C6   1  --     80    00    00    1  C -1 Note Off              
 000025C7   1  --     80    48    00    1  C  5 Note Off              

Adding in some serial.print statements rather than sending to USB shows the data is all there in message as expected, but then we can infer that from the fact the sendNoteOn works fine...

I'm really not sure where to go next - does anyone have any ideas?  I could break out all the different MIDI messages, but feel like I ought to work out why the original isn't working for me...

Oh, and a manual MIDICoreUSB.sendNoteOn(60, 127, 1); works absolutely fine too.

(I've not tried the other direction yet - i.e. USB to Arduino to DIN)

Thanks,
Kevin

**Comment by diyelectromusic** [comment]
By the way, adding a second MIDICoreUSB.send(message) after the sendNoteOn/Off just gives me an additional "blank" message too, see this output from MidiOx for a single keypress and release:

 TIMESTAMP IN PORT STATUS DATA1 DATA2 CHAN NOTE EVENT               
 00001DE9   1  --     90    00    00    1  C -1 Note Off              
 00001DE9   1  --     90    45    7F    1  A  4 Note On               
 00001DE9   1  --     90    00    00    1  C -1 Note Off              
 00002663   1  --     80    00    00    1  C -1 Note Off              
 00002663   1  --     80    45    00    1  A  4 Note Off              
 00002663   1  --     80    00    00    1  C -1 Note Off              

I've just deleted and reinstalled the MIDI library, USB-MIDI and MIDIUSB and it is still the same...

**Reply by paul-emile-element** [reply]
I had the same problem. I identified some issues with the base midi library (not in this repo). I have it working with some changes. I need to start a discussion in that repo.

I had issues with sysex messages and real time messages (active sensing in my case).

Keep in mind that you may have to increase the message size if you want to use sysex.

You may also have to turn thru off. Otherwise, everything will get echoed back to the source.
 MIDICoreUSB.turnThruOff();
 MIDICoreSerial.turnThruOff();
 
https://github.com/FortySevenEffects/arduino_midi_library/blob/master/src/MIDI.hpp

**Reply by diyelectromusic** [reply]
Oh, that's interesting.  What was the issue?  I've used the base MIDI library quite a lot with various routing options between Serial MIDI and USB device MIDI, but I've not seen this issue before...

**Reply by paul-emile-element** [reply]
I joined the discussion and added details here:
https://github.com/FortySevenEffects/arduino_midi_library/issues/159

**Comment by Petrus125** [comment]
I'm also currently working over this DIN2USB sketch. In fact, I would like to get the oposite, USB2Din, and seems should work, but does not. I tested:
-DIN2 USB: Works fine ( for notes, cc etc)
-Din2Din: works fine.
-USB2Din not working
-USB2USB not working

Tested with 32U4 bds. Leonardo, ProMicro.

Wish that USB2Din also work, but could not get. Any one tested or know how to ?

**Reply by paul-emile-element** [reply]
I just tested the example code again on a pro micro and it works for me. I have a midi synth connected to the din in/out so l still have the midi echo problem which can be fixed by turning thru off - but that must be done after the begin:

void setup()
{
  MIDICoreUSB.setHandleMessage(onUsbMessage);
  MIDICoreSerial.setHandleMessage(onSerialMessage);

  MIDICoreUSB.begin(MIDI_CHANNEL_OMNI);
  MIDICoreSerial.begin(MIDI_CHANNEL_OMNI);

  MIDICoreUSB.turnThruOff();
  MIDICoreSerial.turnThruOff();
}

**Reply by Petrus125** [reply]
Are you sure? just only plug the DinOut from ProMicro to DinIn from synth. Not both Dins.
Sure echoing from USB to din?

**Reply by paul-emile-element** [reply]
I actually have five of them that I use as MIDI interfaces. I have been using them for over a year.
![20221105_220907](https://user-images.githubusercontent.com/43148870/200150887-1499506e-ea4a-4c24-817e-1d2a57cd17a8.jpg)

**Reply by Petrus125** [reply]
Hi Paul,wow I see, and did you do any special mod to the original DIN2USB code so to get USB2DinOut?

#include <USB-MIDI.h>
USING_NAMESPACE_MIDI;
typedef USBMIDI_NAMESPACE::usbMidiTransport __umt;
typedef MIDI_NAMESPACE::MidiInterface<__umt> __ss;
__umt usbMIDI(0); // cableNr
__ss MIDICoreUSB((__umt&)usbMIDI);

typedef Message<MIDI_NAMESPACE::DefaultSettings::SysExMaxSize> MidiMessage;

MIDI_CREATE_INSTANCE(HardwareSerial, Serial1, MIDICoreSerial);
// -----------------------------------------------------------------------------
void setup()
{
  MIDICoreUSB.setHandleMessage(onUsbMessage);
  MIDICoreSerial.setHandleMessage(onSerialMessage);

  MIDICoreUSB.begin(MIDI_CHANNEL_OMNI);
  MIDICoreSerial.begin(MIDI_CHANNEL_OMNI);
//  MIDICoreUSB.turnThruOff();
//  MIDICoreSerial.turnThruOff();
}
// -----------------------------------------------------------------------------
void loop()
{
  MIDICoreUSB.read();
  MIDICoreSerial.read();
}
// -----------------------------------------------------------------------------
void onUsbMessage(const MidiMessage& message)
{
  MIDICoreSerial.send(message);
}
void onSerialMessage(const MidiMessage& message)
{
  MIDICoreUSB.send(message);
}

**Reply by paul-emile-element** [reply]
The code should work as is and provide bidirectional communication.
If you need to transmit sysex messages, you might need to adjust the sysex buffer size.

**Reply by Petrus125** [reply]
Something is out of my understanding..... I could not get usb2Din at all.  Tried all option, no way. 
Din2USB perfect. Using original leonardo  with standard  ports Din Shield. App to test: MacPocket Midi.

![IMG_6316](https://user-images.githubusercontent.com/112938746/200201539-a359eca0-0533-4194-bae1-f8aae63c26af.jpg)

**Reply by paul-emile-element** [reply]
Do you see any activity on the tx/rx LEDs on the Leonardo board?

**Reply by paul-emile-element** [reply]
I did clean test with a fresh install on a different computer and the code did not work. I need to figure out what the difference is. I might have some uncommitted code changes in my local repo.

**Comment by Petrus125** [comment]
Yesssss! Aha!  Also came to same conclusion after 1000 tests..... So your running code has to be different.
Conclusion:
-DIN2 USB: Works fine ( for notes, cc etc)
-USB2Din not working. But.......the Din midi In interface light shows led In  activity. Means message is arrived, but not understood.

Changing parameters to your added parameters 
// MIDICoreUSB.turnThruOff();
// MIDICoreSerial.turnThruOff();
makes huge changes with strange results. Finally got USB2 Din data flow, but was receiving lots of shitty data.
Seem like USB2Din is based over a kind of strange Thru ghost.

Alternate, with exactly same setup, tried other libraries examples, and surprise... all bidirectional flow were perfect.
So, the winner is.......: something is weird with the Din2 Usb example ( related to Leonardo? ProMicro? 32U4 caused?) dunno.
Problem for me is now that I have based all my project over this library, and now all needs to be redone.....

**Reply by paul-emile-element** [reply]
I don't have any local change or fix, but I have been working off repos and not with the official releases.
There seems to be some coupling with the MIDI_Library https://github.com/FortySevenEffects/arduino_midi_library. So I need to be on the master branch of both repos for the example to work.

**Reply by Petrus125** [reply]
Still working over this issue.Would be great if you can find the version that is working for you, so to compare with the official one. Sure has to be simple to fix. Still have access to these off repos?

**Reply by paul-emile-element** [reply]
USB-MIDI (master)
$ git log
commit ab62611b9ef50a32875910d4af74eea260bbdcde (HEAD -> master, origin/master, origin/HEAD)
Author: lathoub <4082369+lathoub@users.noreply.github.com>
Date:   Sat Mar 26 23:18:31 2022 +0100

    Update README.md

MIDIUSB (master)
$ git log
commit 140e9f351180c447e7b183197e4a6a71e5e4fde1 (HEAD -> master, origin/master, origin/HEAD)
Merge: c2e5b0d a09f99d
Author: Karl Söderby <35461661+karlsoderby@users.noreply.github.com>
Date:   Fri Oct 1 16:12:05 2021 +0200

    Merge pull request #75 from arduino-libraries/karlsoderby/fix-docs

    Update api.md

MIDI_Library (master)
$ git log
commit 1944aca891df74be1a989bafba5fe618e4d9bd91 (HEAD -> master, origin/master, origin/HEAD)
Author: François Best <francois@francoisbest.com>
Date:   Wed Nov 3 15:06:11 2021 +0100

    chore: Add @paul-emile-element as contributor

**Reply by paul-emile-element** [reply]
Any luck?

**Reply by Petrus125** [reply]
Hi Paul, thank you very much for this info. I was using the latest master branches previously.
Once I checked out this specific commits it worked perfect! :)
So, full tested: ( Leonardo bd)
-DIN2 USB: Works fine ( for notes, cc etc)
-Din2Din: works fine.
-USB2Din: perfect now!

**Reply by paul-emile-element** [reply]
Great! I am happy to hear that this finally works for you!
